### DISCUSSION TITLE
Business Rules

### DISCUSSION THREAD

**Discussion by Luka803** [MarimerLLC/csla#3775] [Questions] [answered]
Hello community!

I have created custom async business rule which should check if the account with the same name already exists which work just fine, it triggers and return proper result.

```
private class NameAlreadyExist : BusinessRuleAsync
{
    private IUnitOfWork UnitOfWork { get; set; }

    public NameAlreadyExist(IUnitOfWork unitOfWork, IPropertyInfo property) : base(property)
    {
        UnitOfWork = unitOfWork;
        IsAsync = true;
        ProvideTargetWhenAsync = true;
    }
    protected override async Task ExecuteAsync(IRuleContext context)
    {
        var target = (AccountEdit)context.Target;
        var input = (string)ReadProperty(target, NameProperty);
        var result = await UnitOfWork.Account.NameAlreadyExist(input);
        if (result)
            context.AddErrorResult($"Account with this name: {input} already exist");
    }

}
```
The problem is that my ```BrokenRulesCollection``` is null so i cant display message from this particular rule. For example, my ``` DataAnnotation ``` rule displays just fine but those custom rules are not provided in ```BrokenRulesCollection```. I ll also provide other code for rules:

1) Calling rule
``` this.BusinessRules.AddRule(new NameAlreadyExist(service, NameProperty)); ```
2) Input field at UI where i mapped NameProperty 
``` 
<div>
    <label>@Property.FriendlyName</label>
    <input @bind-value="@TextValue" @bind-value:event="oninput" />
    <span class="text-danger">@Property.ErrorText</span>
    <span>Is busy: @Property.IsBusy</span>
</div>

@code {
#nullable disable
    [Parameter]
    public Csla.Blazor.IPropertyInfo Property { get; set; }

    private string TextValue
    {
        get => (string)Property.Value;
        set => Property.Value = value;
    }
}
```

Thanks in advance!

**Comment by ajohnstone-ks** [comment]
I'm not sure why BRC would be null, but I would ask why you're using the Target and not `context.GetInputProperty`.  Also your UOW can't be injected like that; you'll need to use `context.ApplicationContext.GetRequiredService<IUnitOfWork>()`  

Is your business rule add code in an override for `AddBusinessRules`

**Reply by Luka803** [reply]
Hey @ajohnstone-ks, thanks for response. 

For first question, i was looking at ProjectTracker project from Git examples and there are few custom rules that are defined like i did. 
I injected UOW at ```AddBusinessRules``` method and then passed that into rule constructor:

```
protected override void AddBusinessRules()
{
    var service = ApplicationContext.GetRequiredService<IUnitOfWork>();
    base.AddBusinessRules();
    this.BusinessRules.AddRule(new NameAlreadyExist(service)
    {
        PrimaryProperty = NameProperty
    });

}
```
I set breakpoint at UI Save method and ``` BrokenRulesColletion ``` of ``` vm.Model ``` is null, which is the whole problem... If you have any ideas, feel free to type. 
Thanks!

**Reply by Luka803** [reply, marked as answer]
If someone would have same problem, it cannot go like i did with UOW. You must go via ``` IDataPortal ``` of the object which will return list of the objects from the database or it's just my soultion.

```
 private class NameAlreadyExist : BusinessRuleAsync
 {
     public NameAlreadyExist(IPropertyInfo property) : base(property)
     {
         ProvideTargetWhenAsync = true;
         IsAsync = true;
         InputProperties = new List<IPropertyInfo> { PrimaryProperty };
     }
     protected override async Task ExecuteAsync(IRuleContext context)
     {
         var service = context.ApplicationContext.GetRequiredService<IDataPortal<AccountList>>();
         string input = (string)context.InputPropertyValues[PrimaryProperty];
         var result = await service.FetchAsync();
         foreach (var item in result)
         {
             if (item.Name == input)
             {
                 context.AddErrorResult($"Account with this input {input} already exist");
                 break;
             }
         }
     }
 }

```
Now, it work just fine.

**Reply by StefanOssendorf** [reply]
I'm glad you solved your problem.
One thought though: If your `AccountList` is growing over time, you are loading the whole list _potentially_ lot of times onto the client just to check if the name is already in use. My suggestion here is to use a dedicated command which something like `NameAlreadyExistCommand` which uses the `Name` as a parameter and returns a `bool`. This saves imho _a lot_ of traffic between client and server.

**Reply by Luka803** [reply]
Hey @StefanOssendorf thank you for the suggestion. I am aware of that, but i am a beginer in term of CSLA, so i still catch up things and the concepts of the framework. This is just example i am practicing on , not a real application 😄. I will consider your solution! 

Thanks!
